### ISSUE TITLE
Discrepancy in Reproducibility of EDM vs. CDM Results on Conditional CIFAR10

### ISSUE THREAD

**Issue by zmtomorrow** [giannisdaras/cdm#5]
Hello,

We are currently working to replicate the results presented in your recent paper, specifically the comparison between EDM and CDM on the conditional CIFAR10 dataset.

Our experiments show that EDM consistently outperforms CDM, contradicting our expectations based on the paper's narrative. Below are the detailed results at various checkpoints:

The first column is the test FID, the last column is the training iterations:
### 00002-cifar10-32x32-cond-ddpmpp-edm-gpus4-batch512-fp32
2.48235 edm network-snapshot-030106
2.19387 edm network-snapshot-050176
1.94555 edm network-snapshot-090317
1.86281 edm network-snapshot-200000

### 00004-cifar10-32x32-cond-ddpmpp-cdm-gpus4-batch512-fp32
2.74859 cdm network-snapshot-030106
2.43844 cdm network-snapshot-050176
2.33994 cdm network-snapshot-057702
2.26956 cdm network-snapshot-070246

We appreciate your work and are eager to accurately reproduce and build upon your findings. Any insights or additional information you could provide would be highly beneficial.

Thank you for your time and consideration.

**Comment by giannisdaras** [comment]
Hi Mingtian, 
thanks a lot for your interest in our work! Something seems off. I would be glad to help you figure it out.
Could you tell me if you are using the pre-trained checkpoints we provide or whether you are training from scratch? 
If you are training from scratch, are you doing so for both the edm and the cdm models? Could you send me the commands you are using to train for both models?

**Comment by J-zin** [comment]
thank you for your super quick reply.

We train both cdm and edm from scrach using your code. here are the commands we use

training edm:
`CUDA_VISIBLE_DEVICES='4' torchrun --standalone --nproc_per_node=1 train.py --outdir=training-runs --batch=512 --batch-gpu=16 --data=/home/zo122/datasets/cifar10-cdm/cifar10-32x32.zip --cond=1 --arch=ddpmpp --precond=edm`

training cdm:
`CUDA_VISIBLE_DEVICES='4' torchrun --standalone --nproc_per_node=1 train.py --outdir=training-runs --batch=512 --batch-gpu=16 --data=/home/zo122/datasets/cifar10-cdm/cifar10-32x32.zip --cond=1 --arch=ddpmpp --precond=cdm`

**Comment by giannisdaras** [comment]
Hm, the only differences I see compared to the command we provide in the repo, i.e. `torchrun --standalone --nproc_per_node=8 train.py --outdir=training-runs \
    --data=datasets/cifar10-32x32.zip --cond=1 --arch=ddpmpp --precond=cdm`, are related to the batch size.
Generally choosing a small batch size could potentially impact the performance of CDM since there is extra stochasticity coming in from the approximation of the expectation in the consistency loss.

If I understand correctly, you still use a batch size of 512, but you are training on a single GPU with 32 gradient accumulation steps?

**Comment by J-zin** [comment]
correct, we use gradient accumulation for cdm and edm

**Comment by giannisdaras** [comment]
Tbh, I have never used gradient accumulation with this codebase, so I am not sure if that's what's causing this.
Just to verify that there is no issue in the FID computation, could you use the given cdm checkpoints and let me know what FID scores you are getting?

**Comment by giannisdaras** [comment]
Btw, doesn't it make it super slow to use 32 gradient accumulation steps? I assume you are limited by your available hardware and you need to use 1 GPU?

**Comment by J-zin** [comment]
> Btw, doesn't it make it super slow to use 32 gradient accumulation steps? I assume you are limited by your available hardware and you need to use 1 GPU?

Sorry for the confusion. We actually use 8 GPUs. The correct command should be

training edm:
`CUDA_VISIBLE_DEVICES='0,1,2,3,4,5,6,7' torchrun --standalone --nproc_per_node=8 train.py --outdir=training-runs --batch=512 --batch-gpu=16 --data=/home/zo122/datasets/cifar10-cdm/cifar10-32x32.zip --cond=1 --arch=ddpmpp --precond=edm`

training cdm:
`CUDA_VISIBLE_DEVICES='0,1,2,3,4,5,6,7' torchrun --standalone --nproc_per_node=8 train.py --outdir=training-runs --batch=512 --batch-gpu=16 --data=/home/zo122/datasets/cifar10-cdm/cifar10-32x32.zip --cond=1 --arch=ddpmpp --precond=cdm`

we use gradient accumulation cuz we train both cdm and edm on RTX 3090.

Will provide u with the FID on your public checkpoints later

**Comment by giannisdaras** [comment]
Ok, this makes more sense. Let me know what numbers you are getting with the public checkpoints. In the meantime, I will try running your exact command and will investigate whether the issue is the gradient accumulation.

**Comment by J-zin** [comment]
Hi,

Here are the FID results tested on the public checkpoints

### CDM
2.57503 cdm network-snapshot-027597.pkl
2.0931 cdm network-snapshot-070246.pkl
1.94711 cdm network-snapshot-100352.pkl
1.92938 cdm network-snapshot-150528.pkl
1.85626 cdm network-snapshot-180634.pkl
1.94957 cdm network-snapshot-200000.pkl

### EDM
2.61678 edm network-snapshot-027597.pkl
2.1003 edm network-snapshot-070246.pkl
2.00657 edm network-snapshot-100352.pkl
1.882 edm network-snapshot-150528.pkl
1.89133 edm network-snapshot-180634.pkl
1.94966 edm network-snapshot-200000.pkl

**Comment by giannisdaras** [comment]
Ok, these numbers make more sense. EDM only outperforms CDM for 150k steps, as we report in the paper. The numbers are slightly inflated here, but that's probably because for the paper we generated 150k images, we computed FID for each one of the 50k splits and we reported the best (page 9).

So, I think the problem has to do with the training, not the evaluation. I will check whether it is the gradient accumulation that's causing it and I will let you know!

**Comment by J-zin** [comment]
Thank you very much for your assistance. If it's not too much trouble, please let me know if the same issue of gradient accumulation happens on your side.

BTW, we found the 1-d toy example on your [poster](https://twitter.com/giannis_daras/status/1734954235081511074/photo/1) very interesting, which can demonstrate the effectiveness of the consistency regularizer. Have you tried to learn the score using cdm on the toy example? If it's possible, we would greatly appreciate it if you could share the code for this toy example with us.

**Comment by zmtomorrow** [comment]
Thank you very much for your support. It's also worth to draw attention to the comparison between the EDM trained by ourselves (based on the commands previously mentioned) and the pre-trained EDM available in the repository:

# EDM trained by ourselves
2.48235 edm network-snapshot-030106
2.19387 edm network-snapshot-050176
1.94555 edm network-snapshot-090317
1.86281 edm network-snapshot-200000

# Provided pre-trained EDM:
2.61678 edm network-snapshot-027597
2.1003 edm network-snapshot-070246
2.00657 edm network-snapshot-100352
1.882 edm network-snapshot-150528
1.89133 edm network-snapshot-180634
1.94966 edm network-snapshot-200000

We are wondering if this is because of the different random seeds used in different runs or if the gradient accumulation training will cause such a discrepancy.  We would appreciate any additional insights you might have on this matter.

It would also be highly beneficial to have some illustrative toy problems that showcase the effectiveness of the consistency regularizer. We would greatly appreciate such examples!

**Comment by zmtomorrow** [comment]
Hi, we are wondering if there are any updates demonstrating the effectiveness of the regularizer on either toy or big problems. 

"We computed FID for each one of the 50k splits and we reported the best."  -- We believe this is not a scientific argument since we can run EDM multiple times and selecting the best outcome could outperform the worst one.   It is necessary to report either the average of multiple runs or the best result from multiple attempts for both EDM and CDM. 
Given that this process is computationally expensive, any toy examples that offer a fair comparison would also be valuable.

Thanks

**Comment by giannisdaras** [comment]
Hey @zmtomorrow.
I would like to kindly ask you to please refrain from provocative public comments such as "we believe this is not a scientific argument". 

We reported the FID the way I mentioned because that's exactly how the EDM paper computes it and we wanted a fair comparison with our baseline. Page 44 from the EDM paper: "To reduce the impact of random variation, [...], 
we compute FID three times in each experiment and report the minimum". We use the same method and we clearly state it in our paper (page 9). We use this method for both EDM and CDM. I would also like to underline that this is an accepted, peer-reviewed work. While it is absolutely fine to have your own opinion about how the evaluation should have been done, I think it is not helpful to publicly judge the work with statements such as "this is not a scientific argument".

Regarding toy examples, we do not have such in the paper, but there are a couple in this very closely related work: https://arxiv.org/abs/2210.04296.

The main issue I think we need to resolve is why you could not replicate the CIFAR-10 numbers. I did not have the time (and the computational resources) to re-run the training with gradient accumulation to see whether this is the issue. I plan to do so at my earliest convenience. If you have the resources to do so, you could also try to train a few times and report the mean and the std of the FID scores you are getting with and without regularization.

**Comment by zmtomorrow** [comment]
Hi, giannisdaras.

Thanks for the reply. We really like the story in the paper and really hope the regularizer can improve the score estimation.

As previously reported, we have trained both CDM and EDM under the same conditions and found that EDM performs better than CDM. Our EDM results are also better than the EDM reported in the repo.   Given the results we have, I can also argue that EDM is better than CDM if we report the best of EDM. That is why we hope this is just a rare case and would really love to see the regularizer could help on average.

Thanks
Mingtian
